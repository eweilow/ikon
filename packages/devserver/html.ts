import { IconGenerationComponent, generateTags } from "@eweilow/ikon";
import assert from "assert";
import Datauri from "datauri/parser";
import { extname } from "path";

type StreamWriter = (chunk: string) => void;

export async function generateTagsHTML(
  Component: IconGenerationComponent,
  write: StreamWriter,
  outDir: string,
  publicPath: string,
  extraBodyContent: string = "",
  concurrency: number = 4
) {
  write(`<html><body><pre>`);
  await generateTags(
    Component,
    publicPath,
    outDir,
    concurrency,
    (tag) => write(tag.replace(/</g, "&lt;").replace(/>/g, "&gt;") + "\n"),
    undefined,
    true
  );
  write(`</pre>`);
  write(extraBodyContent);
  write(`</body></html>`);
}

export async function generateIconsHTML(
  Component: IconGenerationComponent,
  write: StreamWriter,
  outDir: string,
  publicPath: string,
  useRealFiles: boolean,
  extraBodyContent: string = "",
  concurrency: number = 4
) {
  write(`<html><body>`);
  await generateTags(
    Component,
    publicPath,
    outDir,
    concurrency,
    undefined,
    (name, image, publicName) => {
      let src!: string;
      if (useRealFiles) {
        src = publicName;
      } else {
        const outputUri = new Datauri();
        outputUri.format(extname(name), image);
        assert(outputUri.content != null);
        src = outputUri.content;
      }

      write(`<img src="${src}"/>\n`);
    },
    true
  );
  write(extraBodyContent);
  write(`</body></html>`);
}
