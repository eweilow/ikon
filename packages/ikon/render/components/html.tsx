import React from "react";

export const Html = (props: React.PropsWithChildren<{ background: string }>) => (
  <html>
    <head>
      <style
        dangerouslySetInnerHTML={{
          __html: `html, body { 
            min-height: 100%; 
            width: 100%; 
            margin: 0; 
            padding: 0; 
            background: ${props.background}; 
            color: #696969 
          }`,
        }}
      />
    </head>
    <body>
      <main>{props.children}</main>
    </body>
  </html>
);
