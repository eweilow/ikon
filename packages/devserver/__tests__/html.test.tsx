import React from "react";
import { DirectoryResult, dir } from "tmp-promise";

import { generateIconsHTML } from "../html";

describe("generateIconsHTML", () => {
  let tmpDir: DirectoryResult;
  jest.setTimeout(60000);

  beforeEach(async () => {
    tmpDir = await dir({ unsafeCleanup: true });
  });

  afterEach(async () => {
    await tmpDir.cleanup();
  });

  it("renders correctly", async () => {
    const file: string[] = [];
    await generateIconsHTML(
      (props) => {
        const r = Math.floor(props.width % 255);
        const g = Math.floor(props.height % 255);
        const b = Math.floor((props.pixelRatio * 100) % 255);
        return (
          <div
            style={{
              height: "100%",
              width: "100%",
              color: "#" + [r, g, b].map((s) => s.toString(16)).join(""),
              background: "#" + [r, g, b].map((s) => (255 - s).toString(16)).join(""),
              textAlign: "center",
              position: "relative",
            }}
          >
            <div
              style={{
                position: "absolute",
                top: "50%",
                left: "50%",
                marginLeft: -8,
                marginTop: -8,
                height: 16,
                width: 16,
                borderRadius: 8,
                background: "white",
              }}
            />
          </div>
        );
      },
      (chunk) => file.push(chunk),
      tmpDir.path,
      "/",
      true,
      "",
      1
    );

    expect(file.join("\n")).toMatchSnapshot();
  });
});
