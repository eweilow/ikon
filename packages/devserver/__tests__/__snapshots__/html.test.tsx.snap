// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateIconsHTML renders correctly 1`] = `
"<html><body>
<img src=\\"//favicon-16-16.png?h=0f737d68\\"/>

<img src=\\"//favicon-24-24.png?h=34b420cf\\"/>

<img src=\\"//favicon-32-32.png?h=60e36d1e\\"/>

<img src=\\"//favicon-48-48.png?h=036a05b7\\"/>

<img src=\\"//favicon-64-64.png?h=930cb892\\"/>

<img src=\\"//favicon-72-72.png?h=3a6978c3\\"/>

<img src=\\"//favicon-96-96.png?h=5c78d3b3\\"/>

<img src=\\"//favicon-128-128.png?h=a4ed9eab\\"/>

<img src=\\"//android-chrome-36-36.png?h=87c84a33\\"/>

<img src=\\"//android-chrome-48-48.png?h=036a05b7\\"/>

<img src=\\"//android-chrome-72-72.png?h=3a6978c3\\"/>

<img src=\\"//android-chrome-96-96.png?h=5c78d3b3\\"/>

<img src=\\"//android-chrome-144-144.png?h=2fd1dee9\\"/>

<img src=\\"//android-chrome-192-192.png?h=b71edbe2\\"/>

<img src=\\"//android-chrome-256-256.png?h=31805e14\\"/>

<img src=\\"//android-chrome-384-384.png?h=687240c7\\"/>

<img src=\\"//android-chrome-512-512.png?h=be9db060\\"/>

<img src=\\"//apple-touch-icon-57-57.png?h=5e91d777\\"/>

<img src=\\"//apple-touch-icon-60-60.png?h=06e9addf\\"/>

<img src=\\"//apple-touch-icon-72-72.png?h=3a6978c3\\"/>

<img src=\\"//apple-touch-icon-76-76.png?h=da8a656d\\"/>

<img src=\\"//apple-touch-icon-114-114.png?h=a25026c0\\"/>

<img src=\\"//apple-touch-icon-120-120.png?h=e170d3e5\\"/>

<img src=\\"//apple-touch-icon-144-144.png?h=2fd1dee9\\"/>

<img src=\\"//apple-touch-icon-152-152.png?h=777fee3d\\"/>

<img src=\\"//apple-touch-icon-167-167.png?h=d7d07ec2\\"/>

<img src=\\"//apple-touch-icon-180-180.png?h=cf34fc39\\"/>

<img src=\\"//apple-touch-icon.png?h=c105e66b\\"/>

<img src=\\"//apple-touch-icon-precomposed.png?h=cf34fc39\\"/>

<img src=\\"//apple-launch-se-640-1136.png?h=6187ea2a\\"/>

<img src=\\"//apple-launch-8-750-1334.png?h=d00d1398\\"/>

<img src=\\"//apple-launch-xs-1125-2436.png?h=26ab1400\\"/>

<img src=\\"//apple-launch-12-1170-2532.png?h=395b698d\\"/>

<img src=\\"//apple-launch-11-828-1792.png?h=df2133dc\\"/>

<img src=\\"//apple-launch-11-pro-max-1242-2688.png?h=9b4fcc40\\"/>

<img src=\\"//apple-launch-8-plus-1242-2208.png?h=b463e068\\"/>

<img src=\\"//apple-launch-6-plus-1242-2208.png?h=b463e068\\"/>

<img src=\\"//apple-launch-12-pro-max-1284-2778.png?h=61bcccc8\\"/>

<img src=\\"//apple-launch-screen-default.png?h=a2c9052b\\"/>

<img src=\\"//apple-launch-ipad-pro-s-1536-2048.png?h=feb86114\\"/>

<img src=\\"//apple-launch-ipad-pro-1620-2160.png?h=91532c76\\"/>

<img src=\\"//apple-launch-ipad-pro-l-1668-2388.png?h=91636ef2\\"/>

<img src=\\"//apple-launch-air-1668-2224.png?h=82cf14f9\\"/>

<img src=\\"//apple-launch-ipad-pro-xl-2048-2732.png?h=394d65bc\\"/>

<img src=\\"//favicon.ico?h=41d290c7\\"/>


</body></html>"
`;
